I tried building `pkgsStatic.rustc` to test the host platform part, but it looks like LLVM would need some fixes first:
```
  /nix/store/w5ywkv7ph4zji8lg46640p39042ajxw2-x86_64-unknown-linux-musl-binutils-2.31.1/bin/x86_64-unknown-linux-musl-ld: /nix/store/jgrasq40yhnvarrscqxyav98773mf5ib-x86_64-unknown-linux-musl-stage-final-gcc-debug-9.3.0/lib/gcc/x86_64-unknown-linux-musl/9.3.0/crtbeginT.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a shared object
  /nix/store/w5ywkv7ph4zji8lg46640p39042ajxw2-x86_64-unknown-linux-musl-binutils-2.31.1/bin/x86_64-unknown-linux-musl-ld: /nix/store/jgrasq40yhnvarrscqxyav98773mf5ib-x86_64-unknown-linux-musl-stage-final-gcc-debug-9.3.0/lib/gcc/x86_64-unknown-linux-musl/9.3.0/crtend.o: relocation R_X86_64_32 against `.ctors' can not be used when making a shared object; recompile with -fPIC
  /nix/store/w5ywkv7ph4zji8lg46640p39042ajxw2-x86_64-unknown-linux-musl-binutils-2.31.1/bin/x86_64-unknown-linux-musl-ld: final link failed: nonrepresentable section on output
  collect2: error: ld returned 1 exit status
  make[2]: *** [tools/polly/lib/CMakeFiles/LLVMPolly.dir/build.make:196: lib/LLVMPolly.so] Error 1
  make[1]: *** [CMakeFiles/Makefile2:31037: tools/polly/lib/CMakeFiles/LLVMPolly.dir/all] Error 2
  make: *** [Makefile:171: all] Error 2
builder for '/nix/store/1dkhrzniv1lvpza62jm6d0q0xbkinzjh-llvm-10.0.1-x86_64-unknown-linux-musl.drv' failed with exit code 1
cannot build derivation '/nix/store/rjswaxfhn3f37cifl2kxd89rr4asrsbv-rustc-1.47.0-x86_64-unknown-linux-musl.drv': 1 dependencies couldn't be built
```

--- a/pkgs/development/compilers/rust/rustc.nix
+++ b/pkgs/development/compilers/rust/rustc.nix
@@ -1,5 +1,4 @@
 { stdenv, removeReferencesTo, pkgsBuildBuild, pkgsBuildHost, pkgsBuildTarget
-, targetPackages
 , fetchurl, file, python3
 , llvm_10, darwin, cmake, rust, rustPlatform
 , pkgconfig, openssl
@@ -93,8 +92,12 @@ in stdenv.mkDerivation rec {
     "${setTarget}.llvm-config=${llvmSharedForTarget}/bin/llvm-config"
   ] ++ optionals (stdenv.isLinux && !stdenv.targetPlatform.isRedox) [
     "--enable-profiler" # build libprofiler_builtins
+  ] ++ optionals stdenv.buildPlatform.isMusl [
+    "${setBuild}.musl-root=${pkgsBuildBuild.targetPackages.stdenv.cc.libc}"
+  ] ++ optionals stdenv.hostPlatform.isMusl [
+    "${setHost}.musl-root=${pkgsBuildHost.targetPackages.stdenv.cc.libc}"
   ] ++ optionals stdenv.targetPlatform.isMusl [
-    "${setTarget}.musl-root=${targetPackages.stdenv.cc.libc}"
+    "${setTarget}.musl-root=${pkgsBuildTarget.targetPackages.stdenv.cc.libc}"
   ];
 
   # The bootstrap.py will generated a Makefile that then executes the build.

